### PR TITLE
add kruise-rollout addons

### PR DESCRIPTION
**Pull request by zmberg**
Signed-off-by: liheng.zms <liheng.zms@alibaba-inc.com>

### Description of your changes
add kruise-rollout addons

### CODE REVIEW

**Comment by msftgits** [comment]
[CLA assistant check](https://cla.opensource.microsoft.com/oam-dev/catalog?pullRequest=328) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/oam-dev/catalog?pullRequest=328) before we can accept your contribution.<br/><br/>:x: zmberg [sign now](https://cla.opensource.microsoft.com/oam-dev/catalog?pullRequest=328)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/oam-dev/catalog?pullRequest=328) it.</sub>

File: experimental/addons/kruise-rollout/template.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    steps:
+      - name: apply-resources
+        type: apply-application
+  components:
+    - name: kruise-rollout
+      type: helm
+      properties:
+        repoType: helm
+        url: https://openkruise.github.io/charts/
+        chart: kruise-rollout
```
Comment (barnettZQG) [inline experimental/addons/kruise-rollout/template.yaml:17]: The controller deployed 2 instances default?

File: experimental/addons/kruise-rollout/definitions/kruise-rollout.cue [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	}
+	#TrafficRouting: {
+		// use context.name as service if not filled
+		service:            *"" | string
+		gracePeriodSeconds: *5 | int
+		type:               "nginx" | "alb"
+		ingress: name: *"" | string
+	}
+	parameter: {
+		auto: *false | bool
```
Comment (barnettZQG) [inline experimental/addons/kruise-rollout/definitions/kruise-rollout.cue:33]: Please add the help message for the field.
```
// +usage=xxxx
```

File: experimental/addons/kruise-rollout/definitions/kruise-rollout.cue [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+}
+template: {
+	#CanaryStep: {
+		weight?:   int
+		replicas?: int | string
+		duration?: int
+	}
+	#TrafficRouting: {
+		// use context.name as service if not filled
+		service:            *"" | string
```
Comment (barnettZQG) [inline experimental/addons/kruise-rollout/definitions/kruise-rollout.cue:27]:
```suggestion
		service?:             string
```

File: experimental/addons/kruise-rollout/definitions/kruise-rollout.cue [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		gracePeriodSeconds: *5 | int
+		type:               "nginx" | "alb"
+		ingress: name: *"" | string
```
Comment (barnettZQG) [inline experimental/addons/kruise-rollout/definitions/kruise-rollout.cue:30]:
```suggestion
		ingress: name?: string
```

**Comment by wangyikewxgm** [comment]
Cause the experimental addon won't be tested by default, please add test like [argocd](https://github.com/oam-dev/catalog/blob/38ed00362197ad6256dd710653922accb614446a/test/e2e-test/hack/addon-vela-test.sh#L31).

**Comment by wonderflow** [comment]
> Cause the experimental addon won't be tested by default, please add test like [argocd](https://github.com/oam-dev/catalog/blob/38ed00362197ad6256dd710653922accb614446a/test/e2e-test/hack/addon-vela-test.sh#L31).

This should be an official addon instead of experimental, right?

**Comment by Somefive** [comment]
```cue
"kruise-rollout": {
	type: "trait"
	annotations: {}
	labels: {}
	description: "Rollout workload by kruise controller."
	attributes: {
		podDisruptive: true
		appliesToWorkloads: ["*"]
		status: {
			customStatus: #"""
				message: context.outputs.rollout.status.message
				"""#
			healthPolicy: #"""
				isHealth: context.outputs.rollout.status.phase == "Healthy"
				"""#
		}
	}
}
template: {
	#CanaryStep: {
		weight?:   int
		replicas?: int | string
		duration?: int
	}
	#TrafficRouting: {
		// use context.name as service if not filled
		service:            *"" | string
		gracePeriodSeconds: *5 | int
		type:               "nginx" | "alb"
		ingress: name: *"" | string
	}
	parameter: {
		auto: *false | bool
		canary: {
			steps: [...#CanaryStep]
			trafficRoutings?: [...#TrafficRouting]
		}
	}

    srcName: context.output.metadata.name

	outputs: rollout: {
		apiVersion: "rollouts.kruise.io/v1alpha1"
		kind:       "Rollout"
		metadata: {
			name:      context.name
			namespace: context.namespace
		}
		spec: {
			objectRef: {
				workloadRef: {
					apiVersion: context.output.apiVersion
					kind:       context.output.kind
                    if srcName != _|_ {
                        name: srcName
                    }
                    if srcName == _|_ {
                        name: context.name
                    }
				}
			}
			strategy: {
				canary: {
					steps: [
						for k, v in parameter.canary.steps {
							if v.weight != _|_ {
								weight: v.weight
							}

							if v.canaryReplicas != _|_ {
								replicas: v.replicas
							}

							pause: {
								if parameter.auto {
									duration: 0
								}
								if !parameter.auto && v.duration != _|_ {
									duration: v.duration
								}
							}
						},
					]
					if parameter.canary.trafficRoutings != _|_ {
						trafficRoutings: [
							for routing in parameter.canary.trafficRoutings {
								if routing.service != "" {
									service: routing.service
								}
								if routing.service == "" {
									service: context.name
								}
								gracePeriodSeconds: routing.gracePeriodSeconds
								type:               routing.type
								ingress: {
									if routing.ingress.name != "" {
										name: routing.ingress.name
									}
									if routing.ingress.name == "" {
										name: context.name
									}
								}
							},
						]
					}
				}
			}
		}
	}
}
```
Fix the kruise-rollout.cue definition so that it can not only support ref-objects typed component but also support webservice typed component.

File: experimental/addons/kruise-rollout/template.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+apiVersion: core.oam.dev/v1beta1
+kind: Application
+metadata:
+  name: kruise-rollout
+  namespace: vela-system
+spec:
+  workflow:
+    steps:
+      - name: apply-resources
+        type: apply-application
```
Comment (wangyikewxgm) [inline experimental/addons/kruise-rollout/template.yaml:10]: There is no need add these lines. Please remove

**Review by wangyikewxgm — CHANGES_REQUESTED** [review]

**Review by wangyikewxgm — DISMISSED** [review]
LGTM

File: experimental/addons/kruise-rollout/definitions/kruise-rollout.cue [hunk 6]
```diff
@@ -0,0 +1,121 @@
+"kruise-rollout": {
+	type: "trait"
+	annotations: {}
+	labels: {}
+	description: "Rollout workload by kruise controller."
+	attributes: {
+		podDisruptive: true
+		appliesToWorkloads: ["*"]
```
Comment (barnettZQG) [inline experimental/addons/kruise-rollout/definitions/kruise-rollout.cue:8]: Please set the appliesToWorkloads, does the rollout only support the Deployment and CloneSet?
```
appliesToWorkloads: ["deployments.apps", "clonesets.apps.kruise.io"]
```

**Review by wangyikewxgm — APPROVED** [review]
LGTM, good job, thanks

**Pull request merged**